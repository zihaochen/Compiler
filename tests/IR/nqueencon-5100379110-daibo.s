.data
	_printf_cnt: .space 4
	str_1: .space 12
	str_2: .space 12
	str_3: .space 12
	str_4: .space 8
	str_5: .space 8
.text
main:
	move $fp, $sp     # start using memory here
	move $gp, $sp     # set global pointer (unused)
	subu $sp, $sp, 4
	li $t0, 0
	sw $t0, _printf_cnt
	li $t0, 32
	la $t1, str_1
	sw $t0, 0($t1)
	li $t0, 46
	la $t1, str_1
	sw $t0, 4($t1)
	li $t0, 0
	la $t1, str_1
	sw $t0, 8($t1)
	li $t0, 32
	la $t1, str_2
	sw $t0, 0($t1)
	li $t0, 79
	la $t1, str_2
	sw $t0, 4($t1)
	li $t0, 0
	la $t1, str_2
	sw $t0, 8($t1)
	li $t0, 32
	la $t1, str_3
	sw $t0, 0($t1)
	li $t0, 46
	la $t1, str_3
	sw $t0, 4($t1)
	li $t0, 0
	la $t1, str_3
	sw $t0, 8($t1)
	li $t0, 10
	la $t1, str_4
	sw $t0, 0($t1)
	li $t0, 0
	la $t1, str_4
	sw $t0, 4($t1)
	li $t0, 10
	la $t1, str_5
	sw $t0, 0($t1)
	li $t0, 0
	la $t1, str_5
	sw $t0, 4($t1)
	j _main
_printrow:
	sw $ra, ($sp)
	la $t0, str_1
	sw $t0, -20($sp)
	la $t0, str_2
	sw $t0, -36($sp)
	la $t0, str_3
	sw $t0, -56($sp)
	la $t0, str_4
	sw $t0, -76($sp)
	li $t0, 1
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -8($sp)
L2:
	li $t2, 1
	lw $t1, 8($sp)
	sub $t0, $t1, $t2
	sw $t0, -12($sp)
	lw $t2, -12($sp)
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -16($sp)
	lw $t0, -16($sp)
	bne $t0, $0, L3
	b L1
L3:
	lw $t0, -20($sp)
	sw $t0, -84($sp)
	li $t0, 1
	sw $t0, _printf_cnt
	subu $sp, $sp, 88
	jal _printf
	addi $sp, $sp, 88
	sw $v0, -24($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -32($sp)
	lw $t0, -32($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -28($sp)
	b L2
L1:
	lw $t0, -36($sp)
	sw $t0, -84($sp)
	li $t0, 1
	sw $t0, _printf_cnt
	subu $sp, $sp, 88
	jal _printf
	addi $sp, $sp, 88
	sw $v0, -40($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, 8($sp)
	add $t0, $t1, $t2
	sw $t0, -48($sp)
	lw $t0, -48($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -44($sp)
L5:
	lw $t2, 4($sp)
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -52($sp)
	lw $t0, -52($sp)
	bne $t0, $0, L6
	b L4
L6:
	lw $t0, -56($sp)
	sw $t0, -84($sp)
	li $t0, 1
	sw $t0, _printf_cnt
	subu $sp, $sp, 88
	jal _printf
	addi $sp, $sp, 88
	sw $v0, -60($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -72($sp)
	lw $t0, -72($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -68($sp)
	li $t2, 1
	lw $t1, -68($sp)
	sub $t0, $t1, $t2
	sw $t0, -64($sp)
	b L5
L4:
	lw $t0, -76($sp)
	sw $t0, -84($sp)
	li $t0, 1
	sw $t0, _printf_cnt
	subu $sp, $sp, 88
	jal _printf
	addi $sp, $sp, 88
	sw $v0, -80($sp)
	lw $ra, ($sp)
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra
_nqueen:
	sw $ra, ($sp)
	la $t0, str_5
	sw $t0, -200($sp)
	li $t0, 0
	sw $t0, -4($sp)
	lw $t0, 4($sp)
	sw $t0, -8($sp)
	li $t2, 2
	lw $t1, 4($sp)
	rem $t0, $t1, $t2
	sw $t0, -16($sp)
	lw $t0, -16($sp)
	sw $t0, -12($sp)
	li $t2, 2
	lw $t1, 4($sp)
	div $t0, $t1, $t2
	sw $t0, -20($sp)
	li $t2, 3
	lw $t1, -20($sp)
	rem $t0, $t1, $t2
	sw $t0, -24($sp)
	li $t2, 1
	lw $t1, -24($sp)
	sne $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t0, -28($sp)
	bne $t0, $0, L8
	b L9
L8:
	li $t0, 2
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -32($sp)
	lw $ra, ($sp)
	li $t0, 4
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -36($sp)
L11:
	lw $t2, 4($sp)
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -40($sp)
	lw $t0, -40($sp)
	bne $t0, $0, L12
	b L10
L12:
	lw $t0, -4($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -44($sp)
	lw $ra, ($sp)
	li $t2, 2
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -52($sp)
	lw $t0, -52($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -48($sp)
	b L11
L10:
	li $t0, 1
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -56($sp)
L14:
	lw $t2, 4($sp)
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -60($sp)
	lw $t0, -60($sp)
	bne $t0, $0, L15
	b L13
L15:
	lw $t0, -4($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -64($sp)
	lw $ra, ($sp)
	li $t2, 2
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -72($sp)
	lw $t0, -72($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -68($sp)
	b L14
L13:
	b L7
L9:
	lw $t2, -12($sp)
	lw $t1, 4($sp)
	sub $t0, $t1, $t2
	sw $t0, -80($sp)
	lw $t0, -80($sp)
	sw $t0, 4($sp)
	lw $t0, 4($sp)
	sw $t0, -76($sp)
	li $t2, 2
	lw $t1, 4($sp)
	div $t0, $t1, $t2
	sw $t0, -84($sp)
	lw $t0, -84($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -88($sp)
	lw $ra, ($sp)
	li $t2, 2
	lw $t1, 4($sp)
	div $t0, $t1, $t2
	sw $t0, -96($sp)
	li $t2, 1
	lw $t1, -96($sp)
	add $t0, $t1, $t2
	sw $t0, -100($sp)
	lw $t0, -100($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -92($sp)
L17:
	li $t2, 2
	lw $t1, 4($sp)
	div $t0, $t1, $t2
	sw $t0, -104($sp)
	li $t2, 1
	lw $t1, -104($sp)
	sub $t0, $t1, $t2
	sw $t0, -108($sp)
	lw $t2, -108($sp)
	lw $t1, -4($sp)
	sne $t0, $t1, $t2
	sw $t0, -112($sp)
	lw $t0, -112($sp)
	bne $t0, $0, L18
	b L16
L18:
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -116($sp)
	lw $t0, -116($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -120($sp)
	lw $ra, ($sp)
	li $t2, 2
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -128($sp)
	lw $t2, 4($sp)
	lw $t1, -128($sp)
	rem $t0, $t1, $t2
	sw $t0, -132($sp)
	lw $t0, -132($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -124($sp)
	b L17
L16:
	li $t2, 2
	lw $t1, -4($sp)
	sub $t0, $t1, $t2
	sw $t0, -140($sp)
	lw $t2, 4($sp)
	lw $t1, -140($sp)
	rem $t0, $t1, $t2
	sw $t0, -144($sp)
	lw $t0, -144($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -136($sp)
L20:
	li $t2, 2
	lw $t1, 4($sp)
	div $t0, $t1, $t2
	sw $t0, -148($sp)
	li $t2, 1
	lw $t1, -148($sp)
	sub $t0, $t1, $t2
	sw $t0, -152($sp)
	lw $t2, -152($sp)
	lw $t1, -4($sp)
	sne $t0, $t1, $t2
	sw $t0, -156($sp)
	lw $t0, -156($sp)
	bne $t0, $0, L21
	b L19
L21:
	lw $t2, -4($sp)
	lw $t1, 4($sp)
	sub $t0, $t1, $t2
	sw $t0, -160($sp)
	lw $t0, -160($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -164($sp)
	lw $ra, ($sp)
	li $t2, 2
	lw $t1, -4($sp)
	sub $t0, $t1, $t2
	sw $t0, -172($sp)
	lw $t2, 4($sp)
	lw $t1, -172($sp)
	add $t0, $t1, $t2
	sw $t0, -176($sp)
	lw $t2, 4($sp)
	lw $t1, -176($sp)
	rem $t0, $t1, $t2
	sw $t0, -180($sp)
	lw $t0, -180($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -168($sp)
	b L20
L19:
	lw $t2, -4($sp)
	lw $t1, 4($sp)
	sub $t0, $t1, $t2
	sw $t0, -184($sp)
	lw $t0, -184($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -188($sp)
	lw $ra, ($sp)
	lw $t0, -12($sp)
	bne $t0, $0, L23
	b L22
L23:
	li $t2, 1
	lw $t1, 4($sp)
	add $t0, $t1, $t2
	sw $t0, -192($sp)
	lw $t0, -192($sp)
	sw $t0, -208($sp)
	lw $t0, -8($sp)
	sw $t0, -212($sp)
	subu $sp, $sp, 216
	jal _printrow
	addi $sp, $sp, 216
	sw $v0, -196($sp)
	lw $ra, ($sp)
L22:
L7:
	lw $t0, -200($sp)
	sw $t0, -208($sp)
	li $t0, 1
	sw $t0, _printf_cnt
	subu $sp, $sp, 212
	jal _printf
	addi $sp, $sp, 212
	sw $v0, -204($sp)
	lw $ra, ($sp)
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra
_main:
	sw $ra, ($sp)
	subu $sp, $sp, 212
	jal _getchar
	addi $sp, $sp, 212
	sw $v0, -12($sp)
	lw $ra, ($sp)
	li $t2, 48
	lw $t1, -12($sp)
	sub $t0, $t1, $t2
	sw $t0, -16($sp)
	lw $t0, -16($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -8($sp)
L25:
	li $t2, 11
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -20($sp)
	lw $t0, -20($sp)
	bne $t0, $0, L26
	b L24
L26:
	lw $t0, -4($sp)
	sw $t0, -40($sp)
	subu $sp, $sp, 44
	jal _nqueen
	addi $sp, $sp, 44
	sw $v0, -24($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -36($sp)
	lw $t0, -36($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -32($sp)
	li $t2, 1
	lw $t1, -32($sp)
	sub $t0, $t1, $t2
	sw $t0, -28($sp)
	b L25
L24:
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra

_printf:
	lw $t0, _printf_cnt
	mul $t0, $t0, 4
	add $t0, $t0, $sp
	lw $a1, ($t0)
	move $a2, $t0
		
_printf_loop:
	lb $a0, 0($a1)
	beq $a0, 0, _printf_end
	addu $a1, $a1, 4
	beq $a0, '%', _printf_fmt
	li $v0, 11
	syscall
	b _printf_loop

_printf_fmt:
	lb $a0, 0($a1)
	addu $a1, $a1, 4
	beq $a0, 'd', _printf_int
	beq $a0, 's', _printf_str
	beq $a0, 'c', _printf_char
	beq $a0, '0', _printf_width
	beq $a0, '.', _printf_width

_printf_int:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 1
	syscall
	b _printf_loop

_printf_str:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 4
	syscall
	b _printf_loop

_printf_char:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 11
	syscall
	b _printf_loop

_printf_width:
	lb $t1, 0($a1)
	addu $a1, $a1, 4
	sub $t1, $t1, '1'
	addu $a1, $a1, 4
	subu $a2, $a2, 4
	lw $t0, 0($a2)
	blt $t0, 10, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000000, _printf_width_tmp
	subu $t1, $t1, 1

_printf_width_tmp:
	li $a0, '0'
	beq $t1, 0, _printf_width_tmp2
	subu $t1, $t1, 1
	li $v0, 11
	syscall
	b _printf_width_tmp

_printf_width_tmp2:
	move $a0, $t0
	li $v0, 1
	syscall
	b _printf_loop

_printf_end:
	jr $ra

_getchar:
	li $v0, 12
	syscall
	jr $ra

_malloc_struct:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	move $t0, $v0
	li $v0, 9
	li $a0, 4
	syscall
	sw $t0, ($v0)
	jr $ra

_malloc:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	jr $ra

