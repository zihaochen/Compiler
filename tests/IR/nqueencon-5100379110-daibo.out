_start: 
_printf_cnt = 0

_malloc: 

_getchar: 

_printf: 

_printrow: 
_i = 1
Temp1 = _i
Label :2	Temp2 = pos SUB 1
Temp3 = _i LE Temp2
If Not Equal to Zero Temp3 Goto Label :3	
Goto label:1
Label :3	BasicParam Temp4
_printf_cnt = 1
Temp5 = call _printf , 1
Temp7 = _i ADD 1
_i = Temp7
Temp6 = _i
Goto label:2
Label :1	BasicParam Temp8
_printf_cnt = 1
Temp9 = call _printf , 1
Temp11 = pos ADD 1
_i = Temp11
Temp10 = _i
Label :5	Temp12 = _i LE c
If Not Equal to Zero Temp12 Goto Label :6	
Goto label:4
Label :6	BasicParam Temp13
_printf_cnt = 1
Temp14 = call _printf , 1
Temp17 = _i ADD 1
_i = Temp17
Temp16 = _i
Temp15 = Temp16 SUB 1
Goto label:5
Label :4	BasicParam Temp18
_printf_cnt = 1
Temp19 = call _printf , 1
Return BasicParam 0

_nqueen: 
_i = 0
_c = n
Temp20 = n MOD 2
_odd = Temp20
Temp21 = n DIV 2
Temp22 = Temp21 MOD 3
Temp23 = Temp22 NE 1
If Not Equal to Zero Temp23 Goto Label :8	
Goto label:9
Label :8	BasicParam 2
BasicParam _c
Temp24 = call _printrow , 2
_i = 4
Temp25 = _i
Label :11	Temp26 = _i LE n
If Not Equal to Zero Temp26 Goto Label :12	
Goto label:10
Label :12	BasicParam _i
BasicParam _c
Temp27 = call _printrow , 2
Temp29 = _i ADD 2
_i = Temp29
Temp28 = _i
Goto label:11
Label :10	_i = 1
Temp30 = _i
Label :14	Temp31 = _i LE n
If Not Equal to Zero Temp31 Goto Label :15	
Goto label:13
Label :15	BasicParam _i
BasicParam _c
Temp32 = call _printrow , 2
Temp34 = _i ADD 2
_i = Temp34
Temp33 = _i
Goto label:14
Label :13	Goto label:7
Label :9	Temp36 = n SUB _odd
n = Temp36
Temp35 = n
Temp37 = n DIV 2
BasicParam Temp37
BasicParam _c
Temp38 = call _printrow , 2
Temp40 = n DIV 2
Temp41 = Temp40 ADD 1
_i = Temp41
Temp39 = _i
Label :17	Temp42 = n DIV 2
Temp43 = Temp42 SUB 1
Temp44 = _i NE Temp43
If Not Equal to Zero Temp44 Goto Label :18	
Goto label:16
Label :18	Temp45 = _i ADD 1
BasicParam Temp45
BasicParam _c
Temp46 = call _printrow , 2
Temp48 = _i ADD 2
Temp49 = Temp48 MOD n
_i = Temp49
Temp47 = _i
Goto label:17
Label :16	Temp51 = _i SUB 2
Temp52 = Temp51 MOD n
_i = Temp52
Temp50 = _i
Label :20	Temp53 = n DIV 2
Temp54 = Temp53 SUB 1
Temp55 = _i NE Temp54
If Not Equal to Zero Temp55 Goto Label :21	
Goto label:19
Label :21	Temp56 = n SUB _i
BasicParam Temp56
BasicParam _c
Temp57 = call _printrow , 2
Temp59 = _i SUB 2
Temp60 = Temp59 ADD n
Temp61 = Temp60 MOD n
_i = Temp61
Temp58 = _i
Goto label:20
Label :19	Temp62 = n SUB _i
BasicParam Temp62
BasicParam _c
Temp63 = call _printrow , 2
If Not Equal to Zero _odd Goto Label :23	
Goto label:22
Label :23	Temp64 = n ADD 1
BasicParam Temp64
BasicParam _c
Temp65 = call _printrow , 2
Label :22	Label :7	BasicParam Temp66
_printf_cnt = 1
Temp67 = call _printf , 1
Return BasicParam 0

_main: 
_n = 6
Temp68 = _n
Label :25	Temp69 = _n LE 11
If Not Equal to Zero Temp69 Goto Label :26	
Goto label:24
Label :26	BasicParam _n
Temp70 = call _nqueen , 1
Temp73 = _n ADD 1
_n = Temp73
Temp72 = _n
Temp71 = Temp72 SUB 1
Goto label:25
Label :24	Return BasicParam 0

